// @amplitude/analytics-browser/lib/cjs/plugins/default-page-view-event-enrichment.js
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultPageViewEventEnrichment = void 0;
var tslib_1 = require("tslib");
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
var analytics_types_1 = require("@amplitude/analytics-types");
var constants_1 = require("../constants");
var eventPropertyMap = {
    page_domain: "".concat(constants_1.DEFAULT_EVENT_PREFIX, " Page Domain"),
    page_location: "".concat(constants_1.DEFAULT_EVENT_PREFIX, " Page Location"),
    page_path: "".concat(constants_1.DEFAULT_EVENT_PREFIX, " Page Path"),
    page_title: "".concat(constants_1.DEFAULT_EVENT_PREFIX, " Page Title"),
    page_url: "".concat(constants_1.DEFAULT_EVENT_PREFIX, " Page URL"),
};
var defaultPageViewEventEnrichment = function () {
    var name = '@amplitude/plugin-default-page-view-event-enrichment-browser';
    var type = analytics_types_1.PluginType.ENRICHMENT;
    var setup = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
        return [2 /*return*/, undefined];
    }); }); };
    var execute = function (event) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            if (event.event_type === constants_1.DEFAULT_PAGE_VIEW_EVENT && event.event_properties) {
                event.event_properties = Object.entries(event.event_properties).reduce(function (acc, _a) {
                    var _b = tslib_1.__read(_a, 2), key = _b[0], value = _b[1];
                    var transformedPropertyName = eventPropertyMap[key];
                    if (transformedPropertyName) {
                        acc[transformedPropertyName] = value;
                    }
                    else {
                        acc[key] = value;
                    }
                    return acc;
                }, {});
            }
            return [2 /*return*/, event];
        });
    }); };
    return {
        name: name,
        type: type,
        setup: setup,
        execute: execute,
    };
};
exports.defaultPageViewEventEnrichment = defaultPageViewEventEnrichment;
